File: src/lua/asyncawait.lua

yield = coroutine.yield

async = function (fn)
	co = coroutine.create(fn)
	return function (...)
		callback = function (...)
			_, fn, args = coroutine.resume(co, unpack(arg))
			if coroutine.status(co) ~= 'dead' then
				table.insert(args, callback)
				fn(unpack(args))
			end
		end
		callback(arg)
	end
end

await = function (fn)
	return function (...)
		return fn, arg
	end
end

File: ruby/asyncawait.rb

async = lambda { |fn|
	return lambda { |*args|
		gen = fn[*args]
		callback = lambda { |response|
			if not gen.nil?
				descriptor = gen.resume(response)
				if not descriptor.nil?
					descriptor[0][*descriptor[1], callback]
				end
			end
		}
		callback[nil]
	}
}

to = lambda { |func, *args|
	return [
		lambda { |*args, callback|
			async[func][*args, callback]
		},
		args
	]
}
